### ISSUE TITLE
Give a warning if a local variable is assigned to but never read

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#76]
The type checker should give a warning if a local variable is only assigned to but never read. 
This is usually a result of  a misspelled variable name.

Perhaps only do this if warnings have been explicitly enabled using a command line option or via some other mechanism.

**Comment by dubesar** [comment]
@JukkaL Do you mean something like this:


``` python3
def foo(x:int) -> int:
    x = 5
    y = 5
    return x
```
,in this y maybe misspelled local variable which is assigned to but never read.

**Comment by JukkaL** [comment]
@dubesar Yeah! Or something like this:

```py
def thing() -> str:
    processor = 'default'
    if cond():
        processsor = 'extra'  # Note misspelling -- extra 's'
    return processor
```

**Comment by sudhirkumar43** [comment]
Hi @JukkaL, 
Its like point the error like  processsor is not defined earlier? Since we don't know whether the two words are pointing the same variable or different.

**Comment by JukkaL** [comment]
@sudhirkumar43 Can you rephrase your question? I wasn't quite sure what you mean exactly.

**Comment by sudhirkumar43** [comment]
@JukkaL 
like From the above Example:-
def thing() -> str:
    processor = 'default'
    if cond():
        processsor = 'extra'  # Note misspelling -- extra 's'
    return processor
here should it raise error like processsor is not defined earlier . Or we should print the error like the spelling error. Since we don't know whether the two words (processsor , processor ) are pointing the same variable or different. It may possible that processsor  is the name of other variable. I think we can do one thing that just check if a word is similar to some predefined word then raise an error. for example, processsor is much similar to processor raise the error "either missplled word  or variable is not defined earlier".

**Comment by JukkaL** [comment]
It would be sufficient to give a message such as `Value assigned to "processsor" is never read`.

**Comment by sudhirkumar43** [comment]
@JukkaL 
But here is a different case processsor is never initialized and then how can we use it?

**Comment by dosisod** [comment]
@JukkaL, I have been dealing with a similar issue with Refurb ([see comment](https://github.com/dosisod/refurb/issues/18#issuecomment-1270584798)), curious what your thoughts are on the following:

This currently passes in Mypy, and this issue proposes that a warning should be emitted:

```python
x = 123  # WARN: value unused
x = 456
```

This, I believe is similar in nature, but gives a type error instead:

```python
x = 123    # value is still unused here
x = "abc"  # but you get a type error here
```

```
main.py:2: error: Incompatible types in assignment (expression has type "str", variable has type "int")
```

Adding `--allow-redefinition` fixes this, but only if you read from `x` first:

```python
x = 123
print(x)
x = "abc"
```

A few thoughts:

* Do you think it is too strict of a requirement that `x` must be read first before a new variable is created? I feel like this behaviour gives you an inconvinent type error, when in reality it should be telling you that the value you used is never read (at least in this instance that is).

* Wouldn't it be more advantageous that enabling `allow_redefinition` will make a new `x` variable whenever `x` is assigned, regardless of whether it has been read or not? If you kept the ref count, you would be able to tell pretty easily if the value had been read or not. This would bring us closer to how SSA works in say LLVM, as opposed to only creating a new variable when the type changes.
